File: SMO/BackRest/CS/BackRest/BackupRestore.cs
/*============================================================================
  File:    BackupRestore.cs

  Summary: Implements a sample SMO backup and restore utility in C#.

  Date:    January 25, 2005
------------------------------------------------------------------------------
  This file is part of the Microsoft SQL Server Code Samples.

  Copyright (C) Microsoft Corporation.  All rights reserved.

  This source code is intended only as a supplement to Microsoft
  Development Tools and/or on-line documentation.  See these other
  materials for detailed information regarding Microsoft code samples.

  THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY
  KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
  IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
  PARTICULAR PURPOSE.
============================================================================*/
#region Using directives

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Windows.Forms;
using System.IO;

// SMO namespaces
using Microsoft.SqlServer.Management.Common;
using Microsoft.SqlServer.Management.Smo;
using Microsoft.SqlServer;
using Microsoft.SqlServer.MessageBox;

#endregion

namespace Microsoft.Samples.SqlServer
{
    partial class BackupRestore : Form
    {
        // Use the Server object to connect to a specific server
        private Server SqlServerSelection;

        public BackupRestore()
        {
            InitializeComponent();
        }

        private void GetBackupFileButton_Click(object sender, EventArgs e)
        {
            if (BackupFileTextBox.Text.Length > 0)
            {
                this.openFileDialog1.InitialDirectory
                    = Path.GetDirectoryName(BackupFileTextBox.Text);
                this.openFileDialog1.FileName
                    = Path.GetFileName(BackupFileTextBox.Text);
            }

            if (this.openFileDialog1.ShowDialog(this) == DialogResult.OK)
            {
                BackupFileTextBox.Text = this.openFileDialog1.FileName;
            }
        }

        private void BackupButton_Click(object sender, EventArgs e)
        {
            Cursor csr = null;
            Backup backup;
            Database db;
            BackupDeviceItem backupDeviceItem;

            try
            {
                csr = this.Cursor;   // Save the old cursor
                this.Cursor = Cursors.WaitCursor;   // Display the waiting cursor

                // Get database object from combobox
                db = (Database)DatabasesComboBox.SelectedItem;

                // Backup a complete database to disk
                // Create a new Backup object instance
                backup = new Backup();

                // Backup database action
                backup.Action = BackupActionType.Database;

                // Set backup description
                backup.BackupSetDescription
                    = string.Format(
                    System.Globalization.CultureInfo.InvariantCulture,
                    Properties.Resources.SampleBackup, db.Name);

                // Set backup name
                backup.BackupSetName
                    = string.Format(
                    System.Globalization.CultureInfo.InvariantCulture,
                    Properties.Resources.BackupSetName, db.Name);

                // Set database name
                backup.Database = db.Name;

                // Create a file backup device
                backupDeviceItem = new BackupDeviceItem(
                    BackupFileTextBox.Text, DeviceType.File);

                // Add a new backup device
                backup.Devices.Add(backupDeviceItem);

                // Only store this backup in the set
                backup.Initialize = true;

                // Set the media name
                backup.MediaName = Properties.Resources.MediaName;

                // Set the media description
                backup.MediaDescription = Properties.Resources.MediaDescription;

                // Notify this program every 5% 
                backup.PercentCompleteNotification = 5;

                // Set the backup file retention to the current server run value
                backup.RetainDays = db.Parent.Configuration.MediaRetention.RunValue;

                // Skip the tape header, because we are writing to a file
                backup.SkipTapeHeader = true;

                // Unload the tape after the backup completes
                backup.UnloadTapeAfter = true;

                // Add event handler to show progress
                backup.PercentComplete += new PercentCompleteEventHandler(
                    this.ProgressEventHandler);

                // Generate and print script.
                ResultsTextBox.AppendText(Properties.Resources.GeneratedScript);
                ScrollToBottom();

                // Scripting here is strictly for text display purposes.
                String script = backup.Script(SqlServerSelection);

                ResultsTextBox.AppendText(script + Environment.NewLine);
                ScrollToBottom();
                ResultsTextBox.AppendText(Properties.Resources.BackingUp);
                ScrollToBottom();
                UpdateStatus(0);

                // Actual backup starts here.
                backup.SqlBackup(SqlServerSelection);

                ResultsTextBox.AppendText(Properties.Resources.BackupComplete);
                ScrollToBottom();
            }
            catch (SmoException ex)
            {
                ExceptionMessageBox emb = new ExceptionMessageBox(ex);
                emb.Show(this);
            }
            finally
            {
                // Restore the original cursor
                this.Cursor = csr;
            }
        }

        private void RestoreButton_Click(object sender, EventArgs e)
        {
            // Restore the complete database to disk
            Cursor csr = null;
            Restore restore;
            Database db;
            BackupDeviceItem backupDeviceItem;

            // Are you sure?  Default to No.
            if (System.Windows.Forms.MessageBox.Show(this,
                string.Format(System.Globalization.CultureInfo.InvariantCulture,
                Properties.Resources.ReallyRestore,
                DatabasesComboBox.Text), this.Text, MessageBoxButtons.YesNo,
                MessageBoxIcon.Question, MessageBoxDefaultButton.Button2,
                0) == DialogResult.No)
            {
                return;
            }

            try
            {
                csr = this.Cursor;   // Save the old cursor
                this.Cursor = Cursors.WaitCursor;   // Display the waiting cursor

                // Get database object from combobox
                db = (Database)DatabasesComboBox.SelectedItem;

                // Create a new Restore object instance
                restore = new Restore();

                // Restore database action
                restore.Action = RestoreActionType.Database;

                // Set database name
                restore.Database = db.Name;

                // Create a file backup device
                backupDeviceItem = new BackupDeviceItem(
                    BackupFileTextBox.Text, DeviceType.File);

                // Add database backup device
                restore.Devices.Add(backupDeviceItem);

                // Notify this program every 5% 
                restore.PercentCompleteNotification = 5;

                // Replace the existing database
                restore.ReplaceDatabase = true;

                // Unload the backup device (tape)
                restore.UnloadTapeAfter = true;

                // add event handler to show progress
                restore.PercentComplete += new PercentCompleteEventHandler(
                    this.ProgressEventHandler);

                // generate and print script
                ResultsTextBox.AppendText(Properties.Resources.GeneratedScript);

                System.Collections.Specialized.StringCollection strColl =
                    restore.Script(SqlServerSelection);

                foreach (string script in strColl)
                {
                    ResultsTextBox.AppendText(script + Environment.NewLine);
                }

                ResultsTextBox.AppendText(Properties.Resources.Restoring);
                UpdateStatus(0);

                // Actual restore begins here
                restore.SqlRestore(SqlServerSelection);
                ResultsTextBox.AppendText(Properties.Resources.RestoreComplete);
            }
            catch (SmoException ex)
            {
                ExceptionMessageBox emb = new ExceptionMessageBox(ex);
                emb.Show(this);
            }
            finally
            {
                // Restore the original cursor
                this.Cursor = csr;
            }
        }

        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1031:DoNotCatchGeneralExceptionTypes")]
        private void BackupRestore_Load(object sender, EventArgs e)
        {
            ServerConnection ServerConn = new ServerConnection();
            ServerConnect scForm;
            DialogResult dr;

            try
            {
                // Display the main window first
                this.Show();
                Application.DoEvents();

                ServerConn = new ServerConnection();
                scForm = new ServerConnect(ServerConn);
                dr = scForm.ShowDialog(this);
                if ((dr == DialogResult.OK) &&
                    (ServerConn.SqlConnectionObject.State == ConnectionState.Open))
                {
                    SqlServerSelection = new Server(ServerConn);
                    if (SqlServerSelection != null)
                    {
                        this.Text = Properties.Resources.AppTitle
                            + SqlServerSelection.Name;

                        // Refresh database list
                        ShowDatabases(true);
                    }
                }
                else
                {
                    this.Close();
                }

                if (SqlServerSelection != null)
                {
                    BackupFileTextBox.Text = SqlServerSelection.Settings.
                        BackupDirectory + @"\SmoDemoBackup.bak";
                }
            }
            catch (Exception ex)
            {
                ExceptionMessageBox emb = new ExceptionMessageBox(ex);
                emb.Show(this);
            }
        }

        private void BackupRestore_FormClosed(object sender, FormClosedEventArgs e)
        {
            if (SqlServerSelection != null)
            {
                if (SqlServerSelection.ConnectionContext.SqlConnectionObject.State
                    == ConnectionState.Open)
                {
                    SqlServerSelection.ConnectionContext.Disconnect();
                }
            }
        }

        void ProgressEventHandler(object sender, PercentCompleteEventArgs e)
        {
            ResultsTextBox.AppendText(Properties.Resources.ProgressCharacter);
            ScrollToBottom();
            UpdateStatus(e.Percent);
        }

        private void UpdateStatus(int pct)
        {
            statusBar1.Text = string.Format(
                System.Globalization.CultureInfo.InvariantCulture,
                Properties.Resources.CompletedPercent, pct);
            statusBar1.Refresh();
        }

        private void ScrollToBottom()
        {
            ResultsTextBox.Select();
            ResultsTextBox.SelectionStart = ResultsTextBox.Text.Length;
            ResultsTextBox.ScrollToCaret();
        }

        private void ShowDatabases(bool selectDefault)
        {
            // Show the current databases on the server
            Cursor csr = null;

            try
            {
                csr = this.Cursor;   // Save the old cursor
                this.Cursor = Cursors.WaitCursor;   // Display the waiting cursor

                // Clear control
                DatabasesComboBox.Items.Clear();

                // Limit the properties returned to just those that we use
                SqlServerSelection.SetDefaultInitFields(typeof(Database),
                    new String[] { "Name", "IsSystemObject", "IsAccessible" });

                // Add database objects to combobox; the default ToString will display the database name
                foreach (Database db in SqlServerSelection.Databases)
                {
                    if (db.IsSystemObject == false && db.IsAccessible == true)
                    {
                        DatabasesComboBox.Items.Add(db);
                    }
                }

                if ((selectDefault == true) &&
                    (DatabasesComboBox.Items.Count > 0))
                {
                    DatabasesComboBox.SelectedIndex = 0;
                }
            }
            catch (SmoException ex)
            {
                ExceptionMessageBox emb = new ExceptionMessageBox(ex);
                emb.Show(this);
            }
            finally
            {
                this.Cursor = csr;  // Restore the original cursor
            }
        }
    }
}

File: SMO/BackRest/VB/BackRest/BackupRestore.vb
'============================================================================
'  File:    BackupRestore.vb 
'
'  Summary: Implements a sample SMO backup and restore utility in VB.NET.
'
'  Date:    January 25, 2005
'------------------------------------------------------------------------------
'  This file is part of the Microsoft SQL Server Code Samples.
'
'  Copyright (C) Microsoft Corporation.  All rights reserved.
'
'  This source code is intended only as a supplement to Microsoft
'  Development Tools and/or on-line documentation.  See these other
'  materials for detailed information regarding Microsoft code samples.
'
'  THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY
'  KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
'  IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
'  PARTICULAR PURPOSE.
'============================================================================

Public Class BackupRestore
    ' Use the Server object to connect to a specific server
    Private SqlServerSelection As Server

    Private Sub BackupRestore_Load(ByVal sender As Object, _
        ByVal e As System.EventArgs) Handles MyBase.Load
        Dim ServerConn As ServerConnection
        Dim scForm As ServerConnect
        Dim dr As DialogResult

        ' Display the main window first
        Me.Show()
        Application.DoEvents()

        ' Load and display the server selection dialog
        ServerConn = New ServerConnection()
        scForm = New ServerConnect(ServerConn)
        dr = scForm.ShowDialog(Me)
        If dr = Windows.Forms.DialogResult.OK _
            AndAlso ServerConn.SqlConnectionObject.State = ConnectionState.Open Then
            SqlServerSelection = New Server(ServerConn)
            If Not (SqlServerSelection Is Nothing) Then
                Me.Text = My.Resources.AppTitle & SqlServerSelection.Name

                ' Refresh database list
                ShowDatabases(True)
            End If
        Else
            Me.Close()
        End If

        If Not SqlServerSelection Is Nothing Then
            BackupFileTextBox.Text _
                = SqlServerSelection.Settings.BackupDirectory _
                & "\SmoDemoBackup.bak"
        End If
    End Sub

    Private Sub BackupRestore_Closed(ByVal sender As Object, ByVal e As System.EventArgs) Handles MyBase.Closed
        If Not (SqlServerSelection Is Nothing) Then
            If SqlServerSelection.ConnectionContext.SqlConnectionObject.State _
                = ConnectionState.Open Then
                SqlServerSelection.ConnectionContext.Disconnect()
            End If
        End If
    End Sub

    Private Sub BackupButton_Click(ByVal sender As Object, _
        ByVal e As System.EventArgs) Handles BackupButton.Click
        Dim csr As Cursor = Nothing
        Dim backup As Backup
        Dim db As Database
        Dim backupDeviceItem As BackupDeviceItem

        Try
            csr = Me.Cursor ' Save the old cursor
            Me.Cursor = Cursors.WaitCursor ' Display the waiting cursor

            ' Get database object from combobox
            db = CType(DatabasesComboBox.SelectedItem, Database)

            ' Backup a complete database to disk

            ' Create a new Backup object instance
            backup = New Backup()

            ' Backup database action
            backup.Action = BackupActionType.Database

            ' Set backup description
            backup.BackupSetDescription = "Sample Backup of " & db.Name

            ' Set backup name
            backup.BackupSetName = db.Name & " Backup"

            ' Set database name
            backup.Database = db.Name

            ' Create a file backup device
            backupDeviceItem = New BackupDeviceItem( _
                BackupFileTextBox.Text, DeviceType.File)

            ' Add a new backup device
            backup.Devices.Add(backupDeviceItem)

            ' Only store this backup in the set
            backup.Initialize = True

            ' Set the media name
            backup.MediaName = "Set 1"

            ' Set the media description
            backup.MediaDescription = "Sample Backup Media Set # 1"

            ' Notify this program every 5% 
            backup.PercentCompleteNotification = 5

            ' Set the backup file retention to the current server run value
            backup.RetainDays = db.Parent.Configuration.MediaRetention.RunValue

            ' Skip the tape header, because we are writing to a file
            backup.SkipTapeHeader = True

            ' Unload the tape after the backup completes
            backup.UnloadTapeAfter = True

            ' Add event handler to show progress
            AddHandler backup.PercentComplete, _
                AddressOf Me.ProgressEventHandler

            ' Generate and print script
            ResultsTextBox.AppendText(My.Resources.GeneratedScript)
            ScrollToBottom()

            ' Scripting here is strictly for text display purposes
            Dim script As String = backup.Script(SqlServerSelection)
            ResultsTextBox.AppendText(script & Environment.NewLine)
            ScrollToBottom()

            ResultsTextBox.AppendText(My.Resources.BackingUp)
            ScrollToBottom()
            UpdateStatus(0)

            backup.SqlBackup(SqlServerSelection)

            ResultsTextBox.AppendText(My.Resources.BackupComplete)
            ScrollToBottom()
        Catch ex As SmoException
            Dim emb As New ExceptionMessageBox(ex)
            emb.Show(Me)
        Finally
            ' Restore the original cursor
            Me.Cursor = csr
        End Try
    End Sub

    Private Sub ProgressEventHandler(ByVal sender As Object, _
        ByVal e As PercentCompleteEventArgs)
        ResultsTextBox.AppendText(My.Resources.ProgressCharacter)
        ScrollToBottom()
        UpdateStatus(e.Percent)
    End Sub

    Private Sub UpdateStatus(ByVal pct As Integer)
        statusBar1.Text = String.Format(System.Globalization.CultureInfo.InvariantCulture, _
            My.Resources.CompletedPercent, pct)
        statusBar1.Refresh()
    End Sub

    ''' <summary>
    ''' Restore the complete database to disk
    ''' </summary>
    ''' <param name="sender"></param>
    ''' <param name="e"></param>
    ''' <remarks></remarks>
    Private Sub RestoreButton_Click(ByVal sender As Object, _
        ByVal e As System.EventArgs) Handles RestoreButton.Click
        Dim csr As Cursor = Nothing
        Dim restore As Restore
        Dim db As Database
        Dim backupDeviceItem As BackupDeviceItem

        ' Are you sure?  Default to No.
        If System.Windows.Forms.MessageBox.Show(Me, _
            String.Format(System.Globalization.CultureInfo.InvariantCulture, _
            My.Resources.ReallyRestore, DatabasesComboBox.Text), _
            Me.Text, MessageBoxButtons.YesNo, MessageBoxIcon.Question, _
            MessageBoxDefaultButton.Button2, 0) = Windows.Forms.DialogResult.No Then
            Return
        End If

        Try
            csr = Me.Cursor ' Save the old cursor
            Me.Cursor = Cursors.WaitCursor ' Display the waiting cursor

            ' Get database object from combobox
            db = CType(DatabasesComboBox.SelectedItem, Database)

            ' Create a new Restore object instance
            restore = New Restore

            ' Restore database action
            restore.Action = RestoreActionType.Database

            ' Set database name
            restore.Database = db.Name

            ' Create a file backup device
            backupDeviceItem = New BackupDeviceItem( _
                BackupFileTextBox.Text, DeviceType.File)

            ' Add database backup device
            restore.Devices.Add(backupDeviceItem)

            ' Notify this program every 5% 
            restore.PercentCompleteNotification = 5

            ' Replace the existing database
            restore.ReplaceDatabase = True

            ' Unload the backup device (tape)
            restore.UnloadTapeAfter = True

            ' add event handler to show progress
            AddHandler restore.PercentComplete, _
                AddressOf Me.ProgressEventHandler

            ' Generate and print script
            ResultsTextBox.AppendText(My.Resources.GeneratedScript)

            Dim strColl As System.Collections.Specialized.StringCollection _
                = restore.Script(SqlServerSelection)
            For Each str As String In strColl
                ResultsTextBox.AppendText(str & Environment.NewLine)
            Next

            ResultsTextBox.AppendText(My.Resources.Restoring)
            UpdateStatus(0)

            restore.SqlRestore(SqlServerSelection)

            ResultsTextBox.AppendText(My.Resources.RestoreComplete)
        Catch ex As SmoException
            Dim emb As New ExceptionMessageBox(ex)
            emb.Show(Me)
        Finally
            ' Restore the original cursor
            Me.Cursor = csr
        End Try
    End Sub

    Private Sub GetBackupFileButton_Click(ByVal sender As Object, _
        ByVal e As System.EventArgs) Handles GetBackupFileButton.Click
        If (BackupFileTextBox.Text.Length > 0) Then
            Me.OpenFileDialog1.InitialDirectory _
                = Path.GetDirectoryName(BackupFileTextBox.Text)
            Me.OpenFileDialog1.FileName = Path.GetFileName(BackupFileTextBox.Text)
        End If

        If (Me.OpenFileDialog1.ShowDialog(Me) = Windows.Forms.DialogResult.OK) Then
            BackupFileTextBox.Text = Me.OpenFileDialog1.FileName
        End If
    End Sub

    Private Sub ScrollToBottom()
        ResultsTextBox.Select()
        ResultsTextBox.SelectionStart = ResultsTextBox.Text.Length
        ResultsTextBox.ScrollToCaret()
    End Sub

    ''' <summary>
    ''' Show the current databases on the server
    ''' </summary>
    ''' <param name="selectDefault"></param>
    ''' <remarks></remarks>
    Private Sub ShowDatabases(ByVal selectDefault As Boolean)
        Dim csr As Cursor = Nothing

        Try
            csr = Me.Cursor ' Save the old cursor
            Me.Cursor = Cursors.WaitCursor ' Display the waiting cursor

            ' Clear control
            DatabasesComboBox.Items.Clear()

            ' Limit the properties returned to just those that we use
            SqlServerSelection.SetDefaultInitFields(GetType(Database), _
                New String() {"Name", "IsSystemObject", "IsAccessible"})

            ' Add database objects to combobox; the default ToString will display the database name
            For Each db As Database In SqlServerSelection.Databases
                If db.IsSystemObject = False AndAlso db.IsAccessible = True Then
                    DatabasesComboBox.Items.Add(db)
                End If
            Next

            If selectDefault = True _
                AndAlso DatabasesComboBox.Items.Count > 0 Then
                DatabasesComboBox.SelectedIndex = 0
            End If
        Catch ex As SmoException
            Dim emb As New ExceptionMessageBox(ex)
            emb.Show(Me)
        Finally
            Me.Cursor = csr ' Restore the original cursor
        End Try
    End Sub
End Class
